### PR TITLE
Added support for swarm service isolation mode

### PR DESCRIPTION
**Pull request by simonferquel**
**- What I did**
Added isolation field on swarm service creation / updates to enable setting hyperv/process isolation per service on Windows
**- How I did it**
Updated swarmkit (PR with isolation in containerspec), added the field in swarm.ContainerSpec and in convert logic, use it in the executor
**- How to verify it**
Missing a test for now (incoming)
**- Description for the changelog**
Isolation mode (default, process, hyperv) can be set on swarm services to bypass host node defaults

~~Depends on #34745, otherwise breaks CLI~~

### SwarmKit changes:

https://github.com/docker/swarmkit/compare/872861d2ae46958af7ead1d5fffb092c73afbaf0...28f91d87bd3f75fd039dbb9be49bfd2381019261

From that list; included in this bump are included:

- api: Use TaskStatus Err field for non-terminal errors https://github.com/docker/swarmkit/pull/2287
- Support the init flag for container spec https://github.com/docker/swarmkit/pull/2386
- [Windows] Support for isolation mode on container spec https://github.com/docker/swarmkit/pull/2342
- Increase gRPC request timeout to 20 seconds for sending snapshots https://github.com/docker/swarmkit/pull/2391

### CODE REVIEW

**Comment by friism** [comment]
Fixes https://github.com/moby/moby/issues/31616

**Comment by thaJeztah** [comment]
@simonferquel looks like you vendored using an older version of `vndr`; can you update to the current version, and run `vndr` again?

```
14:45:45 The result of vndr differs
14:45:45 
14:45:45 ?? vendor/github.com/docker/distribution/vendor.conf
14:45:45 ?? vendor/github.com/docker/libnetwork/vendor.conf
14:45:45 ?? vendor/github.com/docker/swarmkit/vendor.conf
14:45:45 ?? vendor/github.com/moby/buildkit/vendor.conf
14:45:45 ?? vendor/github.com/opencontainers/runc/vendor.conf
14:45:45 
14:45:45 Please vendor your package with github.com/LK4D4/vndr.
14:45:45 
```

Also, this needs an update of the swagger.yml, and the API version history; https://github.com/moby/moby/blob/master/docs/api/version-history.md

**Comment by simonferquel** [comment]
Swarm test suite is not executed on Windows. Can't write a usefull test on it. I'll just make a test putting an explicit "default" isolation mode, and will inspect service and container to check if it is not empty.

File: api/swagger.yaml [hunk 1]
```diff
@@ -2260,7 +2260,9 @@ definitions:
                     ConfigName is the name of the config that this references, but this is just provided for
                     lookup/display purposes. The config in the reference will be identified by its ID.
                   type: "string"
-
+          Isolation:
```
Comment (thaJeztah) [inline api/swagger.yaml]: Shouldn't this go into the `TaskTemplate.ContainerSpec`?

File: api/swagger.yaml [hunk 2]
```diff
@@ -2260,7 +2260,9 @@ definitions:
...
+            description: "Isolation mode of the containers running the service (default, process, hyperv..."
+            type: "string"
```
Comment (thaJeztah) [inline api/swagger.yaml]: Perhaps add an enum - what is the behaviour if `""` is passed? Is that the same as `default`?

```
            enum:
              - "default"
              - "process"
              - "hyperv"
```

File: api/swagger.yaml [hunk 3]
```diff
@@ -2260,7 +2260,9 @@ definitions:
                     ConfigName is the name of the config that this references, but this is just provided for
                     lookup/display purposes. The config in the reference will be identified by its ID.
                   type: "string"
-
+          Isolation:
+            description: "Isolation mode of the containers running the service (default, process, hyperv..."
```
Comment (thaJeztah) [inline api/swagger.yaml]: Perhaps the same terminology as is used on `docker run`, and add a note that it's only used for Windows containers;

```
description: |
  Container isolation technology to use for the container.

  <p><br /></p>

  > **Note**: Container isolation is only used for Windows containers.
  > This field is ignored for Linux containers.
```

(the `<p><br /></p>` is a bit hacky, but I haven't found another way to add spacing yet :smile:)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left some notes

File: api/swagger.yaml (hunk 2 above)
Comment (simonferquel) [inline api/swagger.yaml]: "" or null have the same behavior as default (except service inspect will report with an explicit "default" value for isolation).
I am not completely certain about making it an enum btw, as underlying executor (hcs for windows) might introduce new isolation mode. I'll check on container/create messages and harmonize that.

**Comment by simonferquel** [comment]
@thaJeztah fixed the doc (used the same exact description and enum as on HostConfig)
We'll need to wait for https://github.com/docker/swarmkit/pull/2342 to be merged, update once again the vendoring before merging this one

**Comment by thaJeztah** [comment]
@simonferquel you forgot this one; https://github.com/moby/moby/pull/34424/commits/967fe0cbceb68a6c25a0ac51a5b368bea29f3a1a#r131908016 😇

File: api/swagger.yaml (hunk 1 above)
Comment (simonferquel) [inline api/swagger.yaml]: It is the case :)

Comment (thaJeztah) [inline api/swagger.yaml]: oh boy, you're right; It looked as if it was at the wrong indentation level, but you're right. Sorry, my mistake ha!

**Comment by thaJeztah** [comment]
@simonferquel this needs a rebase

**Comment by coolljt0725** [comment]
ping @simonferquel needs refactor

**Comment by simonferquel** [comment]
@thaJeztah, could you please update your review ?

File: docs/api/version-history.md [hunk 4]
```diff
@@ -20,6 +20,7 @@ keywords: "API, Docker, rcli, REST, documentation"
 * `POST /containers/create` now accepts additional values for the
   `HostConfig.IpcMode` property. New values are `private`, `shareable`,
   and `none`.
+* `POST /services/create` and `POST /services/(id)/update` now accept an `Isolation` field on container spec
```
Comment (thaJeztah) [inline docs/api/version-history.md]: This should now be in v1.33 as v1.32 has shipped in a release; I opened https://github.com/moby/moby/pull/34908 to bump the API version

File: integration-cli/daemon/daemon.go [hunk 5]
```diff
@@ -678,6 +679,23 @@ func (d *Daemon) FindContainerIP(id string) (string, error) {
 	return strings.Trim(out, " \r\n'"), nil
 }
 
+// InspectContainer inspects a container
+func (d *Daemon) InspectContainer(ctx context.Context, id string) (*container.Config, *container.HostConfig, error) {
```
Comment (thaJeztah) [inline integration-cli/daemon/daemon.go]: These tests now probably have to be moved to the `/integration` suite, not `/integration-cli/`; https://github.com/moby/moby/blob/master/integration/service/create_test.go

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left some comments 😅

**Comment by PatrickLang** [comment]
Can we revive this? With Windows Server version 1709 around the corner, it will be needed to run containers using the Windows Server 2016 base images. Windows versions have breaking kernel changes between them but `--isolation=hyperv` works around that by using the right kernel allowing a newer node to run older containers.

**Comment by thaJeztah** [comment]
@PatrickLang this is waiting for https://github.com/docker/swarmkit/pull/2342 in SwarmKit upstream

**Comment by thaJeztah** [comment]
It also involves an API change so cannot be backported without introducing a lot of problems

**Comment by simonferquel** [comment]
Since the PR Swarmkit side has been merged, I rebased and refreshed this PR.
However, Isolation field on swarm side has moved from a free form string to an enum, with an explicit "Default" value, so I can't use the fact that I could have both <empty>  and "default" valid strings to make a Linux integration kit.

What I want to do then, is to introduce more unit testing, that would hopefully cover all aspects of the PR

**Comment by simonferquel** [comment]
Done (made a bit of cleanup for reusing Isolation type instead of a free string, and added tests to make sure conversions and executors take the change into account)

**Comment by simonferquel** [comment]
I just rebased it to see if the plugin tests still fail. Overwise I think the needs/vendoring flag can be removed (the pr now references swarmkit master)

**Comment by simonferquel** [comment]
Hmm swarmkit vendoring update seem to have broken `DockerSwarmSuite.TestSwarmVolumePlugin` somehow. Not sure why, but I could use some help here...

**Comment by simonferquel** [comment]
Ok, found the reason of the test regression: error details are reported in a separate task status field now. Updated the test, amended my commit, waiting for test results.

**Comment by simonferquel** [comment]
@thaJeztah could you update your review ? Want to make sure I did update the right swagger stuff.

**Review by thaJeztah — APPROVED** [review]
LGTM, but needs a minor rebase

**Comment by thaJeztah** [comment]
ping @cpuguy83 @dnephin PTAL

**Comment by vieux** [comment]
LGTM ping @cpuguy83 @dnephin

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Pull request merged**